src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1830,7 +1836,7 @@ void PerformKeepAlive(object? state)
                 Log.Error("Keepalive failure", e, Id);
                 try
                 {
-                    Break(e);
+                    Break(new NpgsqlException("Connection is broken due to keepalive failure, check logs for more information."));
```

Comment:
BTW it may also be possible to fix this simply by using interlocked operations on the connector's state, without requiring a lock.

Reply:
> BTW it may also be possible to fix this simply by using interlocked operations on the connector's state, without requiring a lock.

Not sure about this one. As I see it:
1. User has an open connection, `KeepAlive` is still performing from time to time.
2. `KeepAlive` sends a request to PG, and is waiting for a response.
3. User started to execute a query, we're waiting in `NpgsqlConnector.StartUserAction` for the `KeepAlive` to return the lock.
4. `KeepAlive` failed due to a timeout, connector's state is broken and connector is cleaned up (this is still while in the lock).
5. `NpgsqlConnector.StartUserAction` is finally free to take a lock, but `_userLock` is `null`, which leads to `NullReferenceException`.

Reply:
Sorry, I made a bit of a mess - the comment on the interlocked was intend for me general review, and only relates to the concurrency detection bit between two different commands being executed on the same connection. It has nothing to do with this Break exception.

Further, even if we do implement concurrency detection via interlocked operations, you're right that we'd still need the current locking strategy between the keepalive and the actual execution, since the execution must block on the keepalive completing.

Bottom line, this PR is fine as-is, once the unrelated comments on the exception itself are fixed.